### PR TITLE
restrict app credentials to 1.19+

### PR DESCRIPTION
**Pull request by kon-angelo**
**How to categorize this PR?**

/area control-plane
/kind bug
/platform openstack

**What this PR does / why we need it**:

Using the [new application credentials](https://github.com/gardener/gardener-extension-provider-openstack/pull/300) is not supported by the in-tree openstack cloud-provider. Currently we use the in-tree provider for up to kubernetes version `v1.18.X`.  

With this PR the validation will block creation of shoots using Openstack's application credentials in unsupported versions. 

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other operator
NONE
```

### CODE REVIEW

**Review by rfranzke — COMMENTED** [review]
Can you add a unit test?

File: pkg/admission/validator/shoot.go [hunk 1]
```diff
@@ -18,20 +18,21 @@ import (
...
 	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"github.com/gardener/gardener/pkg/utils/version"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/equality"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/serializer"
 	"k8s.io/apimachinery/pkg/util/validation/field"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	api "github.com/gardener/gardener-extension-provider-openstack/pkg/apis/openstack"
```
Comment (dkistner) [inline pkg/admission/validator/shoot.go:33]: Nit: Can we move those imports from this repo up?

File: pkg/admission/validator/shoot.go [hunk 2]
```diff
@@ -99,6 +100,17 @@ func (s *shoot) Validate(ctx context.Context, new, old client.Object) error {
...
 	}
 
+	// The Kubernetes version is the version of the CSI migration, where we stopped using the in-tree providers.
+	// see: pkg/webhook/controlplane/ensurer.go
+	k8sVersionLessThan19, err := version.CompareVersions(shoot.Spec.Kubernetes.Version, "<", "1.19")
+	if err != nil {
+		return err
+	}
+
+	if k8sVersionLessThan19 && s.isUsingApplicationCredentials(secret) {
```
Comment (dkistner) [inline pkg/admission/validator/shoot.go:110]: Should we not rather check this in `ExtractCredentials()`?
And pass `k8sVersionLessThan19` then to `ExtractCredentials()`.
Then we could also adapt the unit test there.

Comment (dkistner) [reply pkg/admission/validator/shoot.go:110]: We agreed not to move it to `ExtractCredentials()` as it is used in different places.

**Comment by kon-angelo** [comment]
Made a second try. If this is not up to standard I can also try making the check in `ExtractCredentials`

File: pkg/openstack/credentials.go [hunk 3]
```diff
@@ -119,3 +119,7 @@ func getRequired(secret *corev1.Secret, key string) (string, error) {
 	}
 	return string(value), nil
 }
+
```
Comment (dkistner) [inline pkg/openstack/credentials.go:122]: Do we need this still be as an exported function? I see it is only used in `pkg/apis/openstack/validation/shoot.go`. Can we move it there or remove the function?
If you want to keep it there then please add a doc string.

Comment (kon-angelo) [reply pkg/openstack/credentials.go]: I inlined the function since it's used in one place only and it's a one-liner and amply commented its usage.

**Review by dkistner — DISMISSED** [review]
/lgtm

**Review by dkistner — APPROVED** [review]
/lgtm

**Pull request merged**